File: android/app/src/main/java/com/jianxing/browser/data/RulesStore.kt
package com.jianxing.browser.data

import android.content.Context
import android.content.SharedPreferences
import com.jianxing.browser.model.BiliConstants
import com.jianxing.browser.model.BilibiliExtensionConfig
import com.jianxing.browser.model.RulesConfig
import com.jianxing.browser.model.SiteGroup
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

class RulesStore(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    @Synchronized
    fun load(): RulesConfig {
        val json = prefs.getString(KEY_RULES, null) ?: return defaultRules()
        return try {
            val config = parseRules(JSONObject(json))
            migrateLegacyPassword(config)
        } catch (_: Exception) {
            defaultRules()
        }
    }

    /** Clear legacy SHA-256 hashes so parent must re-set with scrypt. */
    private fun migrateLegacyPassword(config: RulesConfig): RulesConfig {
        val hash = config.parentPasswordHash
        if (hash.isBlank() || hash.startsWith("scrypt$")) return config
        val isLegacyHex = hash.matches(Regex("^[a-fA-F0-9]{64}$"))
        if (!isLegacyHex && !hash.startsWith("scrypt$")) {
            // Unknown format — clear
            val cleared = config.copy(parentPasswordHash = "")
            save(cleared)
            return cleared
        }
        if (isLegacyHex) {
            val cleared = config.copy(parentPasswordHash = "")
            save(cleared)
            return cleared
        }
        return config
    }

    @Synchronized
    fun save(config: RulesConfig) {
        prefs.edit().putString(KEY_RULES, toJson(config).toString()).apply()
    }

    fun hasPassword(): Boolean = load().parentPasswordHash.isNotBlank()

    fun setPassword(password: String) {
        val rules = load()
        save(rules.copy(parentPasswordHash = PasswordHasher.hash(password)))
    }

    fun verifyPassword(password: String): Boolean {
        val hash = load().parentPasswordHash
        return PasswordHasher.verify(password, hash)
    }

    fun changePassword(current: String, next: String): Boolean {
        if (!verifyPassword(current)) return false
        setPassword(next)
        return true
    }

    // ── Group CRUD (Electron parity) ──

    @Synchronized
    fun createGroup(
        name: String,
        extensionId: String = "none",
        useSuggestedHosts: Boolean = true
    ): SiteGroup? {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) return null
        val ext = if (extensionId == "bilibili") "bilibili" else "none"
        val hosts = if (ext == "bilibili" && useSuggestedHosts) {
            BiliConstants.SUGGESTED_HOSTS.toList()
        } else {
            emptyList()
        }
        val group = SiteGroup(
            id = newId(),
            name = trimmed,
            enabled = true,
            hosts = hosts,
            extensionId = ext,
            extensionConfig = defaultExtensionConfig(ext)
        )
        val rules = load()
        save(rules.copy(groups = rules.groups + group))
        return group
    }

    @Synchronized
    fun updateGroup(
        id: String,
        name: String? = null,
        enabled: Boolean? = null,
        extensionId: String? = null
    ): Boolean {
        val rules = load()
        val idx = rules.groups.indexOfFirst { it.id == id }
        if (idx < 0) return false
        var g = rules.groups[idx]
        if (name != null && name.trim().isNotEmpty()) {
            g = g.copy(name = name.trim())
        }
        if (enabled != null) {
            g = g.copy(enabled = enabled)
        }
        if (extensionId == "bilibili" || extensionId == "none") {
            if (extensionId != g.extensionId) {
                g = g.copy(
                    extensionId = extensionId,
                    extensionConfig = defaultExtensionConfig(extensionId)
                )
            }
        }
        val groups = rules.groups.toMutableList()
        groups[idx] = g
        save(rules.copy(groups = groups))
        return true
    }

    @Synchronized
    fun deleteGroup(id: String): Boolean {
        val rules = load()
        val next = rules.groups.filterNot { it.id == id }
        if (next.size == rules.groups.size) return false
        save(rules.copy(groups = next))
        return true
    }

    @Synchronized
    fun setEnabled(id: String, enabled: Boolean): Boolean =
        updateGroup(id, enabled = enabled)

    fun isFilteringEnabled(): Boolean = load().filteringEnabled

    @Synchronized
    fun setFilteringEnabled(enabled: Boolean) {
        save(load().copy(filteringEnabled = enabled))
    }

    fun getHomepage(): String = load().homepage

    @Synchronized
    fun setHomepage(raw: String): Boolean {
        val parsed = normalizeHomepage(raw) ?: return false
        save(load().copy(homepage = parsed))
        return true
    }

    @Synchronized
    fun addHost(groupId: String, host: String): Boolean {
        val cleaned = cleanHost(host)
        if (cleaned.isEmpty()) return false
        val rules = load()
        val idx = rules.groups.indexOfFirst { it.id == groupId }
        if (idx < 0) return false
        val group = rules.groups[idx]
        if (group.hosts.any { it.equals(cleaned, ignoreCase = true) }) return true
        val groups = rules.groups.toMutableList()
        groups[idx] = group.copy(hosts = group.hosts + cleaned)
        save(rules.copy(groups = groups))
        return true
    }

    @Synchronized
    fun removeHost(groupId: String, host: String): Boolean {
        val cleaned = cleanHost(host)
        val rules = load()
        val idx = rules.groups.indexOfFirst { it.id == groupId }
        if (idx < 0) return false
        val group = rules.groups[idx]
        val groups = rules.groups.toMutableList()
        groups[idx] = group.copy(hosts = group.hosts.filterNot { it.equals(cleaned, ignoreCase = true) })
        save(rules.copy(groups = groups))
        return true
    }

    @Synchronized
    fun addBiliUp(groupId: String, mid: String, note: String? = null): Boolean {
        if (!mid.matches(Regex("^\\d+$"))) return false
        val rules = load()
        val idx = rules.groups.indexOfFirst { it.id == groupId }
        if (idx < 0) return false
        val group = rules.groups[idx]
        if (group.extensionId != "bilibili") return false
        val cfg = asBiliConfig(group.extensionConfig)
        val mids = if (cfg.allowedMids.contains(mid)) cfg.allowedMids else cfg.allowedMids + mid
        val notes = cfg.midNotes.toMutableMap()
        if (!note.isNullOrBlank()) notes[mid] = note
        val ext = mutableMapOf<String, Any?>(
            "allowedMids" to mids,
            "midNotes" to notes
        )
        val groups = rules.groups.toMutableList()
        groups[idx] = group.copy(extensionConfig = ext)
        save(rules.copy(groups = groups))
        return true
    }

    @Synchronized
    fun removeBiliUp(groupId: String, mid: String): Boolean {
        val rules = load()
        val idx = rules.groups.indexOfFirst { it.id == groupId }
        if (idx < 0) return false
        val group = rules.groups[idx]
        if (group.extensionId != "bilibili") return false
        val cfg = asBiliConfig(group.extensionConfig)
        val mids = cfg.allowedMids.filterNot { it == mid }
        val notes = cfg.midNotes.toMutableMap().apply { remove(mid) }
        val ext = mutableMapOf<String, Any?>(
            "allowedMids" to mids,
            "midNotes" to notes
        )
        val groups = rules.groups.toMutableList()
        groups[idx] = group.copy(extensionConfig = ext)
        save(rules.copy(groups = groups))
        return true
    }

    fun exportGroups(): List<SiteGroup> =
        load().groups.map { g ->
            g.copy(
                hosts = g.hosts.toList(),
                extensionConfig = g.extensionConfig.toMap()
            )
        }

    fun getGroup(id: String): SiteGroup? = load().groups.find { it.id == id }

    // ── Request-group helpers (kept) ──

    fun ensureDefaultRequestGroup() {
        val rules = load()
        val existing = rules.groups.find {
            it.name == BiliConstants.REQUEST_GROUP_NAME && it.extensionId == "none"
        }
        if (existing != null) return
        createGroup(
            name = BiliConstants.REQUEST_GROUP_NAME,
            extensionId = "none",
            useSuggestedHosts = false
        )
    }

    fun getOrCreateRequestGroup(): SiteGroup {
        ensureDefaultRequestGroup()
        return load().groups.first {
            it.name == BiliConstants.REQUEST_GROUP_NAME && it.extensionId == "none"
        }
    }

    fun addHostToRequestGroup(host: String): Boolean {
        val group = getOrCreateRequestGroup()
        return addHost(group.id, host)
    }

    fun removeHostFromRequestGroup(host: String) {
        val group = load().groups.find {
            it.name == BiliConstants.REQUEST_GROUP_NAME && it.extensionId == "none"
        } ?: return
        removeHost(group.id, host)
    }

    fun replaceGroups(groups: List<SiteGroup>) {
        val rules = load()
        val merged = groups.toMutableList()
        if (merged.none { it.name == BiliConstants.REQUEST_GROUP_NAME && it.extensionId == "none" }) {
            val local = rules.groups.find {
                it.name == BiliConstants.REQUEST_GROUP_NAME && it.extensionId == "none"
            }
            if (local != null) merged.add(local)
            else merged.add(
                SiteGroup(
                    id = newId(),
                    name = BiliConstants.REQUEST_GROUP_NAME,
                    enabled = true,
                    hosts = emptyList(),
                    extensionId = "none"
                )
            )
        }
        save(rules.copy(groups = merged))
    }

    companion object {
        private const val PREFS_NAME = "jianxing_rules"
        private const val KEY_RULES = "rules_json"

        fun newId(): String = "g_${UUID.randomUUID().toString().take(12)}"

        fun cleanHost(host: String): String {
            return host.trim().lowercase()
                .removePrefix("https://")
                .removePrefix("http://")
                .substringBefore('/')
                .trimEnd('.')
        }

        fun defaultExtensionConfig(extensionId: String): Map<String, Any?> {
            return if (extensionId == "bilibili") {
                mapOf(
                    "allowedMids" to emptyList<String>(),
                    "midNotes" to emptyMap<String, String>()
                )
            } else {
                emptyMap()
            }
        }

        fun asBiliConfig(raw: Map<String, Any?>?): BilibiliExtensionConfig {
            val midsRaw = raw?.get("allowedMids")
            val mids = when (midsRaw) {
                is List<*> -> midsRaw.map { it.toString() }.filter { it.matches(Regex("^\\d+$")) }
                is JSONArray -> (0 until midsRaw.length()).map { midsRaw.getString(it) }
                    .filter { it.matches(Regex("^\\d+$")) }
                else -> emptyList()
            }
            val notesRaw = raw?.get("midNotes")
            val notes = mutableMapOf<String, String>()
            when (notesRaw) {
                is Map<*, *> -> notesRaw.forEach { (k, v) ->
                    if (k != null && v != null) notes[k.toString()] = v.toString()
                }
                is JSONObject -> notesRaw.keys().forEach { k ->
                    notes[k] = notesRaw.optString(k, "")
                }
            }
            return BilibiliExtensionConfig(allowedMids = mids, midNotes = notes)
        }

        fun normalizeHomepage(raw: String): String? {
            val s = raw.trim()
            if (s.isEmpty()) return ""
            var url = s
            if (!Regex("^[a-zA-Z][a-zA-Z0-9+.-]*:").containsMatchIn(url)) {
                url = "https://$url"
            }
            return try {
                val uri = android.net.Uri.parse(url)
                val scheme = uri.scheme?.lowercase()
                if (scheme != "http" && scheme != "https") null
                else uri.toString()
            } catch (_: Exception) {
                null
            }
        }

        private fun defaultRules() = RulesConfig(
            version = 2,
            parentPasswordHash = "",
            filteringEnabled = false,
            homepage = "",
            groups = emptyList()
        )

        private fun parseRules(obj: JSONObject): RulesConfig {
            val groupsArr = obj.optJSONArray("groups") ?: JSONArray()
            val groups = mutableListOf<SiteGroup>()
            for (i in 0 until groupsArr.length()) {
                normalizeGroup(groupsArr.getJSONObject(i))?.let { groups.add(it) }
            }
            return RulesConfig(
                version = obj.optInt("version", 2),
                parentPasswordHash = obj.optString("parentPasswordHash", ""),
                filteringEnabled = obj.optBoolean("filteringEnabled", false),
                homepage = normalizeHomepage(obj.optString("homepage", "")) ?: "",
                groups = groups
            )
        }

        private fun normalizeGroup(obj: JSONObject): SiteGroup? {
            val extensionId = if (obj.optString("extensionId") == "bilibili") "bilibili" else "none"
            val hostsArr = obj.optJSONArray("hosts") ?: JSONArray()
            val hosts = (0 until hostsArr.length())
                .map { cleanHost(hostsArr.getString(it)) }
                .filter { it.isNotEmpty() }
                .distinct()
            val extObj = obj.optJSONObject("extensionConfig")
            val extMap = mutableMapOf<String, Any?>()
            if (extObj != null) {
                extObj.keys().forEach { key ->
                    extMap[key] = extObj.get(key)
                }
            }
            if (extensionId == "bilibili" && !extMap.containsKey("allowedMids")) {
                extMap["allowedMids"] = emptyList<String>()
                extMap["midNotes"] = emptyMap<String, String>()
            }
            return SiteGroup(
                id = obj.optString("id").ifBlank { newId() },
                name = obj.optString("name").ifBlank { "未命名配置组" },
                enabled = obj.optBoolean("enabled", true),
                hosts = hosts,
                extensionId = extensionId,
                extensionConfig = extMap
            )
        }

        fun toJson(config: RulesConfig): JSONObject {
            val groups = JSONArray()
            config.groups.forEach { g ->
                groups.put(groupToJson(g))
            }
            return JSONObject()
                .put("version", config.version)
                .put("parentPasswordHash", config.parentPasswordHash)
                .put("filteringEnabled", config.filteringEnabled)
                .put("homepage", config.homepage)
                .put("groups", groups)
        }

        fun groupToJson(g: SiteGroup): JSONObject {
            val hosts = JSONArray()
            g.hosts.forEach { hosts.put(it) }
            val ext = JSONObject()
            g.extensionConfig.forEach { (k, v) ->
                when (v) {
                    is List<*> -> {
                        val arr = JSONArray()
                        v.forEach { arr.put(it) }
                        ext.put(k, arr)
                    }
                    is Map<*, *> -> {
                        val o = JSONObject()
                        v.forEach { (mk, mv) -> o.put(mk.toString(), mv) }
                        ext.put(k, o)
                    }
                    is JSONArray -> ext.put(k, v)
                    is JSONObject -> ext.put(k, v)
                    null -> {}
                    else -> ext.put(k, v)
                }
            }
            return JSONObject()
                .put("id", g.id)
                .put("name", g.name)
                .put("enabled", g.enabled)
                .put("hosts", hosts)
                .put("extensionId", g.extensionId)
                .put("extensionConfig", ext)
        }

        fun parseGroupsFromJson(arr: JSONArray): List<SiteGroup> {
            val list = mutableListOf<SiteGroup>()
            for (i in 0 until arr.length()) {
                normalizeGroup(arr.getJSONObject(i))?.let { list.add(it) }
            }
            return list
        }
    }
}

File: src/main/rules-store.ts
import { app } from 'electron';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs';
import { dirname, join } from 'path';
import { randomBytes } from 'crypto';
import {
  asBiliConfig,
  BILI_SUGGESTED_HOSTS,
  emptyBiliConfig,
  EXTENSION_CATALOG,
  ExtensionId,
  PublicRules,
  RulesConfig,
  SiteGroup,
} from '../shared/types';
import { hashPassword, verifyPassword } from './parent-auth';

function rulesPath(): string {
  return join(app.getPath('userData'), 'rules.json');
}

function newId(): string {
  return `g_${Date.now().toString(36)}_${randomBytes(3).toString('hex')}`;
}

function cleanHost(host: string): string {
  return host
    .trim()
    .toLowerCase()
    .replace(/^https?:\/\//, '')
    .split('/')[0]
    .replace(/\/$/, '');
}

function defaultExtensionConfig(extensionId: ExtensionId): Record<string, unknown> {
  if (extensionId === 'bilibili') {
    return { ...emptyBiliConfig() };
  }
  return {};
}

function defaultRules(): RulesConfig {
  return {
    version: 2,
    parentPasswordHash: '',
    filteringEnabled: false,
    homepage: '',
    groups: [],
  };
}

function readHomepage(raw: unknown): string {
  const parsed = normalizeHomepage(typeof raw === 'string' ? raw : '');
  return parsed.ok ? parsed.url : '';
}

export function normalizeHomepage(raw: string): { ok: true; url: string } | { ok: false; error: string } {
  const s = String(raw || '').trim();
  if (!s) return { ok: true, url: '' };
  let url = s;
  if (!/^[a-zA-Z][a-zA-Z0-9+.-]*:/.test(url)) {
    url = `https://${url}`;
  }
  try {
    const u = new URL(url);
    if (u.protocol !== 'http:' && u.protocol !== 'https:') {
      return { ok: false, error: '仅支持 http/https 地址' };
    }
    return { ok: true, url: u.toString() };
  } catch {
    return { ok: false, error: '网址无效' };
  }
}

function normalizeGroup(raw: Partial<SiteGroup>): SiteGroup | null {
  if (!raw || typeof raw !== 'object') return null;
  const extensionId: ExtensionId =
    raw.extensionId === 'bilibili' ? 'bilibili' : 'none';
  const hosts = Array.isArray(raw.hosts)
    ? [...new Set(raw.hosts.map(cleanHost).filter(Boolean))]
    : [];
  let extensionConfig: Record<string, unknown> =
    raw.extensionConfig && typeof raw.extensionConfig === 'object'
      ? { ...raw.extensionConfig }
      : defaultExtensionConfig(extensionId);
  if (extensionId === 'bilibili') {
    extensionConfig = { ...asBiliConfig(extensionConfig) };
  }
  return {
    id: typeof raw.id === 'string' && raw.id ? raw.id : newId(),
    name: typeof raw.name === 'string' && raw.name.trim() ? raw.name.trim() : '未命名配置组',
    enabled: raw.enabled !== false,
    hosts,
    extensionId,
    extensionConfig,
  };
}

/** Migrate v1 flat whitelist → v2 groups. */
function migrateV1(raw: Record<string, unknown>): RulesConfig {
  const hosts = Array.isArray(raw.allowedHosts)
    ? (raw.allowedHosts as string[]).map(cleanHost).filter(Boolean)
    : [...BILI_SUGGESTED_HOSTS];
  const mids = Array.isArray(raw.allowedBiliMids)
    ? (raw.allowedBiliMids as unknown[]).map(String)
    : [];
  const notes =
    raw.biliUpNotes && typeof raw.biliUpNotes === 'object'
      ? (raw.biliUpNotes as Record<string, string>)
      : {};
  const group: SiteGroup = {
    id: newId(),
    name: 'B 站',
    enabled: raw.biliHostsEnabled !== false,
    hosts: hosts.length ? hosts : [...BILI_SUGGESTED_HOSTS],
    extensionId: 'bilibili',
    extensionConfig: {
      allowedMids: mids,
      midNotes: notes,
    },
  };
  return {
    version: 2,
    parentPasswordHash:
      typeof raw.parentPasswordHash === 'string' ? raw.parentPasswordHash : '',
    filteringEnabled: raw.filteringEnabled === true,
    homepage: '',
    groups: [group],
  };
}

export class RulesStore {
  private rules: RulesConfig;

  constructor() {
    this.rules = this.load();
  }

  private load(): RulesConfig {
    const path = rulesPath();
    if (!existsSync(path)) {
      const initial = defaultRules();
      this.persist(initial);
      return initial;
    }
    try {
      const raw = JSON.parse(readFileSync(path, 'utf8')) as Record<string, unknown>;
      if (raw.version === 2 && Array.isArray(raw.groups)) {
        const groups = (raw.groups as Partial<SiteGroup>[])
          .map(normalizeGroup)
          .filter((g): g is SiteGroup => Boolean(g));
        return {
          version: 2,
          parentPasswordHash:
            typeof raw.parentPasswordHash === 'string'
              ? raw.parentPasswordHash
              : '',
          filteringEnabled: raw.filteringEnabled === true,
          homepage: readHomepage(raw.homepage),
          groups,
        };
      }
      // v1 or unknown → migrate
      const migrated = migrateV1(raw);
      this.persist(migrated);
      return migrated;
    } catch {
      return defaultRules();
    }
  }

  private persist(rules: RulesConfig = this.rules): void {
    const path = rulesPath();
    mkdirSync(dirname(path), { recursive: true });
    writeFileSync(path, JSON.stringify(rules, null, 2), 'utf8');
  }

  getPublic(): PublicRules {
    return {
      hasPassword: Boolean(this.rules.parentPasswordHash),
      filteringEnabled: this.isFilteringEnabled(),
      homepage: this.getHomepage(),
      groups: this.rules.groups.map((g) => ({
        ...g,
        hosts: [...g.hosts],
        extensionConfig: { ...g.extensionConfig },
      })),
      extensions: EXTENSION_CATALOG.map((e) => ({ ...e })),
    };
  }

  getRaw(): RulesConfig {
    return this.rules;
  }

  hasPassword(): boolean {
    return Boolean(this.rules.parentPasswordHash);
  }

  isFilteringEnabled(): boolean {
    return this.rules.filteringEnabled === true;
  }

  getHomepage(): string {
    return this.rules.homepage || '';
  }

  setHomepage(raw: string): { ok: boolean; error?: string; rules?: PublicRules } {
    const parsed = normalizeHomepage(raw);
    if (!parsed.ok) return parsed;
    this.rules.homepage = parsed.url;
    this.persist();
    return { ok: true, rules: this.getPublic() };
  }

  setFilteringEnabled(
    enabled: boolean
  ): { ok: boolean; rules?: PublicRules } {
    this.rules.filteringEnabled = Boolean(enabled);
    this.persist();
    return { ok: true, rules: this.getPublic() };
  }

  setPassword(password: string): { ok: boolean; error?: string } {
    if (!password || password.length < 4) {
      return { ok: false, error: '密码至少 4 位' };
    }
    this.rules.parentPasswordHash = hashPassword(password);
    this.persist();
    return { ok: true };
  }

  changePassword(
    current: string,
    next: string
  ): { ok: boolean; error?: string } {
    if (!this.verify(current)) {
      return { ok: false, error: '当前密码错误' };
    }
    return this.setPassword(next);
  }

  verify(password: string): boolean {
    if (!this.rules.parentPasswordHash) return false;
    return verifyPassword(password, this.rules.parentPasswordHash);
  }

  private findGroup(id: string): SiteGroup | undefined {
    return this.rules.groups.find((g) => g.id === id);
  }

  createGroup(input: {
    name: string;
    extensionId?: ExtensionId;
    useSuggestedHosts?: boolean;
  }): { ok: boolean; error?: string; group?: SiteGroup; rules?: PublicRules } {
    const name = (input.name || '').trim();
    if (!name) return { ok: false, error: '请填写配置组名称' };
    const extensionId: ExtensionId =
      input.extensionId === 'bilibili' ? 'bilibili' : 'none';
    const hosts =
      extensionId === 'bilibili' && input.useSuggestedHosts !== false
        ? [...BILI_SUGGESTED_HOSTS]
        : [];
    const group: SiteGroup = {
      id: newId(),
      name,
      enabled: true,
      hosts,
      extensionId,
      extensionConfig: defaultExtensionConfig(extensionId),
    };
    this.rules.groups.push(group);
    this.persist();
    return { ok: true, group, rules: this.getPublic() };
  }

  updateGroup(
    id: string,
    patch: Partial<Pick<SiteGroup, 'name' | 'enabled' | 'extensionId'>>
  ): { ok: boolean; error?: string; rules?: PublicRules } {
    const group = this.findGroup(id);
    if (!group) return { ok: false, error: '配置组不存在' };
    if (typeof patch.name === 'string' && patch.name.trim()) {
      group.name = patch.name.trim();
    }
    if (typeof patch.enabled === 'boolean') {
      group.enabled = patch.enabled;
    }
    if (patch.extensionId === 'bilibili' || patch.extensionId === 'none') {
      if (patch.extensionId !== group.extensionId) {
        group.extensionId = patch.extensionId;
        group.extensionConfig = defaultExtensionConfig(patch.extensionId);
      }
    }
    this.persist();
    return { ok: true, rules: this.getPublic() };
  }

  deleteGroup(id: string): { ok: boolean; error?: string; rules?: PublicRules } {
    const before = this.rules.groups.length;
    this.rules.groups = this.rules.groups.filter((g) => g.id !== id);
    if (this.rules.groups.length === before) {
      return { ok: false, error: '配置组不存在' };
    }
    this.persist();
    return { ok: true, rules: this.getPublic() };
  }

  addHost(
    groupId: string,
    host: string
  ): { ok: boolean; error?: string; rules?: PublicRules } {
    const group = this.findGroup(groupId);
    if (!group) return { ok: false, error: '配置组不存在' };
    const cleaned = cleanHost(host);
    if (!cleaned) return { ok: false, error: '无效域名' };
    if (!group.hosts.includes(cleaned)) group.hosts.push(cleaned);
    this.persist();
    return { ok: true, rules: this.getPublic() };
  }

  removeHost(
    groupId: string,
    host: string
  ): { ok: boolean; error?: string; rules?: PublicRules } {
    const group = this.findGroup(groupId);
    if (!group) return { ok: false, error: '配置组不存在' };
    group.hosts = group.hosts.filter((h) => h !== host);
    this.persist();
    return { ok: true, rules: this.getPublic() };
  }

  addBiliUp(
    groupId: string,
    mid: string,
    note?: string
  ): { ok: boolean; error?: string; rules?: PublicRules } {
    const group = this.findGroup(groupId);
    if (!group) return { ok: false, error: '配置组不存在' };
    if (group.extensionId !== 'bilibili') {
      return { ok: false, error: '该配置组未启用 B 站扩展' };
    }
    const cfg = asBiliConfig(group.extensionConfig);
    if (!cfg.allowedMids.includes(mid)) cfg.allowedMids.push(mid);
    if (note) cfg.midNotes[mid] = note;
    group.extensionConfig = { ...cfg };
    this.persist();
    return { ok: true, rules: this.getPublic() };
  }

  removeBiliUp(
    groupId: string,
    mid: string
  ): { ok: boolean; error?: string; rules?: PublicRules } {
    const group = this.findGroup(groupId);
    if (!group) return { ok: false, error: '配置组不存在' };
    if (group.extensionId !== 'bilibili') {
      return { ok: false, error: '该配置组未启用 B 站扩展' };
    }
    const cfg = asBiliConfig(group.extensionConfig);
    cfg.allowedMids = cfg.allowedMids.filter((m) => m !== mid);
    delete cfg.midNotes[mid];
    group.extensionConfig = { ...cfg };
    this.persist();
    return { ok: true, rules: this.getPublic() };
  }

  /** Replace all groups from cloud sync (keeps local parent password). */
  replaceGroups(
    groups: SiteGroup[]
  ): { ok: boolean; error?: string; rules?: PublicRules } {
    if (!Array.isArray(groups)) {
      return { ok: false, error: '配置格式无效' };
    }
    const normalized = groups
      .map((g) => normalizeGroup(g))
      .filter((g): g is SiteGroup => Boolean(g));
    this.rules.groups = normalized;
    this.persist();
    return { ok: true, rules: this.getPublic() };
  }

  exportGroups(): SiteGroup[] {
    return this.rules.groups.map((g) => ({
      ...g,
      hosts: [...g.hosts],
      extensionConfig: { ...g.extensionConfig },
    }));
  }
}

export function extractMidFromInput(input: string): string | null {
  const trimmed = input.trim();
  if (/^\d+$/.test(trimmed)) return trimmed;
  try {
    const url = new URL(
      trimmed.startsWith('http') ? trimmed : `https://${trimmed}`
    );
    if (url.hostname.includes('space.bilibili.com')) {
      const m = url.pathname.match(/^\/(\d+)/);
      if (m) return m[1];
    }
    const midParam = url.searchParams.get('mid');
    if (midParam && /^\d+$/.test(midParam)) return midParam;
  } catch {
    // ignore
  }
  const spaceMatch = trimmed.match(/space\.bilibili\.com\/(\d+)/);
  if (spaceMatch) return spaceMatch[1];
  return null;
}
